test/src/unit-cppapi-query-condition-enumerations.cc

```diff
@@ -0,0 +1,1169 @@
+/**
+ * @file unit-cppapi-query-condition-enumerations.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2024 TileDB Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Tests the C++ API for query condition related functions.
+ */
+
+#include <limits>
+#include <ostream>
+#include <random>
+
+#include "test/support/src/ast_helpers.h"
+#include "test/support/tdb_catch.h"
+#include "tiledb/sm/c_api/tiledb_struct_def.h"
+#include "tiledb/sm/cpp_api/tiledb"
+#include "tiledb/sm/cpp_api/tiledb_experimental"
+#include "tiledb/sm/serialization/query.h"
+
+#ifdef TILEDB_SERIALIZATION
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#define GENERATE_SERIALIZATION() GENERATE(false, true)
+#else
+#define GENERATE_SERIALIZATION() false
+#endif
+
+using namespace tiledb;
+
+/*
+ * The test fixture. See the first test for a basic example of expected
+ * usage.
+ */
+
+struct EnmrQCCell;
+struct ResultEnmrQCCell;
+
+using EnmrQCMatcher = std::function<bool(const EnmrQCCell& cell)>;
+using EnmrQCCreator = std::function<QueryCondition(Context& ctx)>;
+
+class CPPQueryConditionEnumerationFx {
+ public:
+  CPPQueryConditionEnumerationFx();
+  ~CPPQueryConditionEnumerationFx();
+
+  uint32_t run_test(
+      tiledb_array_type_t type,
+      bool serialize,
+      EnmrQCMatcher matcher,
+      EnmrQCCreator creator,
+      uint32_t num_rows = 1024);
+
+ protected:
+  std::string uri_;
+  Context ctx_;
+  VFS vfs_;
+
+  tiledb_array_type_t type_;
+  bool serialize_;
+  uint32_t num_rows_;
+
+  // A fill value result. This is the value that a dense array query returns
+  // for a non-matching result.
+  std::unique_ptr<EnmrQCCell> fill_;
+
+  // Our random source
+  std::mt19937_64 rand_;
```

Comment:
Use PRNG?

Reply:
PRNG doesn't expose access to the underlying `std::mt19937_64` instance which is required by the `uniforn_{int|reall}_distribution` interface.